src/tools/compiletest/src/read2.rs

```diff
@@ -3,31 +3,54 @@
 
 pub use self::imp::read2;
 use std::io;
-use std::process::{Child, Output};
+use std::process::{Child, ExitStatus};
 
-pub fn read2_abbreviated(mut child: Child) -> io::Result<Output> {
-    use io::Write;
+pub struct ReadOutput {
+    pub status: ExitStatus,
+    pub stdout: String,
+    pub stderr: String,
+}
+
+pub fn read2_abbreviated(mut child: Child, exclude_from_len: &[String]) -> io::Result<ReadOutput> {
     use std::mem::replace;
 
     const HEAD_LEN: usize = 160 * 1024;
     const TAIL_LEN: usize = 256 * 1024;
 
     enum ProcOutput {
-        Full(Vec<u8>),
+        Full { string: String, excluded_len: usize },
         Abbreviated { head: Vec<u8>, skipped: usize, tail: Box<[u8]> },
     }
 
     impl ProcOutput {
-        fn extend(&mut self, data: &[u8]) {
+        fn extend(&mut self, data: &[u8], exclude_from_len: &[String]) {
             let new_self = match *self {
-                ProcOutput::Full(ref mut bytes) => {
-                    bytes.extend_from_slice(data);
-                    let new_len = bytes.len();
-                    if new_len <= HEAD_LEN + TAIL_LEN {
+                ProcOutput::Full { ref mut string, ref mut excluded_len } => {
+                    let utf8 = String::from_utf8_lossy(data);
+                    string.push_str(&*utf8);
+
+                    // We had problems in the past with tests failing only in some environments,
+                    // due to the length of the base path pushing the output size over the limit.
+                    //
+                    // To make those failures deterministic across all environments we ignore known
+                    // paths when calculating the string length, while still including the full
+                    // path in the output. This could result in some output being larger than the
+                    // threshold, but it's better than having nondeterministic failures.
```

Comment:
#This also defeats the OOM protection, if we have for some reason rustc being over eager with *just* outputting paths. (Or repeating a path). It also won't avoid non-determinism in terms of failures here, at least not those caused by other differences in behavior (e.g., cfg(unix) vs not differences if testing on Windows).

Right now we abbreviate output after 416 KB of output from a single test. It seems like OOM-wise we should be comfortable bumping to, say, 5 megabytes -- I would expect any "reasonable" output from rustc to fit into that for our tests. That should make path length differences less of an issue, IMO, and is a much simpler change and feels less scary.

If we want to keep an eye on the tests so we catch cases where we dump kilobytes into the stderr for no good reason, I think our best bet would be a system closer to perf.rust-lang.org or toolstate; I could pretty easily imagine us dumping each test's output size (and maybe runtime?) into a file and then having a simple comparison post-merge that we're not increasing more than 10% (or staying in some <512KB range or something). I'm not really convinced we actually need to do anything here though.

Reply:
> This also defeats the OOM protection, if we have for some reason rustc being over eager with just outputting paths. (Or repeating a path).

It would have to output just the source or build or source path repeatedly, not a subpath thereof. Otherwise the subpaths would still be counted and eventually exceed the size limit.

Reply:
> This also defeats the OOM protection, if we have for some reason rustc being over eager with just outputting paths. (Or repeating a path).

Pushed a tentative second commit addressing this, by adding a known placeholder of 32 bytes for each excluded pattern. Changing the placeholder to 64 triggers the abbreviation btw, so it seems like a good enough value to keep the behavior close to the current one.

> It also won't avoid non-determinism in terms of failures here, at least not those caused by other differences in behavior (e.g., cfg(unix) vs not differences if testing on Windows).

Sure! It would still prevent some non-determinism, and other kinds like the ones you mention would likely be catched by the CI builders we have.

> Right now we abbreviate output after 416 KB of output from a single test. It seems like OOM-wise we should be comfortable bumping to, say, 5 megabytes -- I would expect any "reasonable" output from rustc to fit into that for our tests. That should make path length differences less of an issue, IMO, and is a much simpler change and feels less scary.

Hmm, while 5MB would likely avoid the issue I'm facing right now, it also wouldn't have catched #94327, which seems like a category of issues worth fixing. Also, the (fairly short) rustfix test in question (`src/test/ui/numeric/numeric-cast.rs`) emits 793 paths in its JSON output, and I could see more complex rustfix tests in the future, emitting way more diagnostics (and thus more paths) than that.

> If we want to keep an eye on the tests so we catch cases where we dump kilobytes into the stderr for no good reason, I think our best bet would be a system closer to perf.rust-lang.org or toolstate; I could pretty easily imagine us dumping each test's output size (and maybe runtime?) into a file and then having a simple comparison post-merge that we're not increasing more than 10% (or staying in some <512KB range or something). I'm not really convinced we actually need to do anything here though.

This seems way more complex than the solution proposed PR, and would also be harder to integrate for someone running CIs outside of the Rust project.

Reply:
I don't think it's necesasry, but now that the safety net is there it doesn't hurt either as long as CI and local builds are equal before this judge.